lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,607 @@
+end
+
+--- Remove one or more widgets from the layout
+-- The last parameter can be a boolean, forcing a recursive seach of the
+-- widget(s) to remove.
+-- @param widget ... Widgets that should be removed (must at least be one)
+-- @treturn boolean If the operation is successful
+function grid:remove_widgets(...)
+    local args = { ... }
+
+    local recursive = type(args[#args]) == "boolean" and args[#args]
+
+    local ret = true
+    for k, rem_widget in ipairs(args) do
+        if recursive and k == #args then break end
+
+        local idx, l = self:index(rem_widget, recursive)
+
+        if idx and l and l.remove then
+            l:remove(idx, false)
+        else
+            ret = false
+        end
+    end
+
+    return #args > (recursive and 1 or 0) and ret
+end
+
+function grid:get_children()
+    return self._private.children
+end
+
+function grid:set_children(children)
+    self:reset()
+    if #children > 0 then
+        self:add(unpack(children))
+    end
+end
+
+--- Replace the first instance of `widget` in the layout with `widget2`
+-- @param widget The widget to replace
+-- @param widget2 The widget to replace `widget` with
+-- @tparam[opt=false] boolean recursive Digg in all compatible layouts to find the widget.
+-- @treturn boolean If the operation is successful
+function grid:replace_widget(widget, widget2, recursive)
+    local idx, l = self:index(widget, recursive)
+
+    if idx and l then
+        l:set(idx, widget2)
+        return true
+    end
+
+    return false
+end
+
+function grid:swap_widgets(widget1, widget2, recursive)
+    base.check_widget(widget1)
+    base.check_widget(widget2)
+
+    local idx1, l1 = self:index(widget1, recursive)
+    local idx2, l2 = self:index(widget2, recursive)
+
+    if idx1 and l1 and idx2 and l2 and (l1.set or l1.set_widget) and (l2.set or l2.set_widget) then
+        if l1.set then
+            l1:set(idx1, widget2)
+        elseif l1.set_widget then
+            l1:set_widget(widget2)
+        end
+        if l2.set then
+            l2:set(idx2, widget1)
+        elseif l2.set_widget then
+            l2:set_widget(widget1)
+        end
+
+        return true
+    end
+
+    return false
+end
+
+function grid:set(index, widget2)
+    if (not index) or (index < 1) or (not widget2) then return false end
+
+    base.check_widget(widget2)
+
+    local max_count = self._private.row_count*self._private.column_count
+
+    if index > max_count then
+        local new_rows_count = math.floor(
+            (index - max_count)/self._private.column_count
+        ) + 1
+
+        self._private.row_count = self._private.row_count + new_rows_count
+
+        for ii=1, self._private.row_count do
+            self._private.rows[ii] = self._private.rows[ii] or {}
+        end
+    end
+
+    -- This ignore if there is a widget or not in the cell
+    local i = math.floor((index - 1) / self._private.column_count) + 1
+    local j = ((index - 1) % self._private.column_count) + 1
```

Comment:
I'm not sure if it is a good idea, but I'll suggest it anyway:
Instead of having both `self._private.children` and `self._private.rows`, how about only having one of them?
Either just `children`, but with `nil`s at cells that are empty (and then having to fix things up if the column count changes), or just having `rows` and... Hm, both ideas make it harder to have a "linear set of indicies".

Ok, how about `local empty_widget = wibox.widget.base.make_widget()` at the beginning and instead of using `nil` for empty cells, this widget is used? This means that the layout "automagically adds entries to itself" when e.g. the row count changes, but it seems like it could simplify the code a lot. The code could then store its children in a "linear table" (`children`) and `i`/`j` can easily be "transformed" into an index into this table.

Anything like this...?